modules/sql/src/test/scala/SqlPaging1Mapping.scala

```diff
@@ -0,0 +1,168 @@
+trait SqlPaging1Mapping[F[_]] extends SqlTestMapping[F] {
+  object root extends RootDef {
+    val numCountries = col("num_countries", int8)
+  }
+
+  object country extends TableDef("country") {
+    val code         = col("code", bpchar(3))
+    val name         = col("name", text)
+    val numCities    = col("num_cities", int8)
+  }
+
+  object city extends TableDef("city") {
+    val id          = col("id", int4)
+    val countrycode = col("countrycode", bpchar(3))
+    val name        = col("name", text)
+  }
+
+  val schema =
+    schema"""
+      type Query {
+        countries(offset: Int!, limit: Int!): PagedCountry!
+      }
+      type PagedCountry {
+        offset: Int!
+        limit: Int!
+        total: Int!
+        items: [Country!]!
+      }
+      type Country {
+        code: String!
+        name: String!
+        cities(offset: Int, limit: Int): PagedCity!
+      }
+      type PagedCity {
+        offset: Int!
+        limit: Int!
+        total: Int!
+        items: [City!]!
+      }
+      type City {
+        name: String!
+      }
+    """
+
+  val QueryType = schema.ref("Query")
+  val PagedCountryType = schema.ref("PagedCountry")
+  val CountryType = schema.ref("Country")
+  val PagedCityType = schema.ref("PagedCity")
+  val CityType = schema.ref("City")
+
+  val typeMappings =
+    List(
+      ObjectMapping(
+        tpe = QueryType,
+        fieldMappings =
+          List(
+            SqlObject("countries"),
+          )
+      ),
+      ObjectMapping(
+        tpe = PagedCountryType,
+        fieldMappings =
+          List(
+            CursorField("offset", genValue("countryOffset"), Nil),
+            CursorField("limit", genValue("countryLimit"), Nil),
+            SqlField("total", root.numCountries),
+            SqlObject("items")
+          )
+      ),
+      ObjectMapping(
+        tpe = CountryType,
+        fieldMappings =
+          List(
+            SqlField("code", country.code, key = true),
+            SqlField("name", country.name),
+            SqlObject("cities")
+          )
+      ),
+      ObjectMapping(
+        tpe = PagedCityType,
+        fieldMappings =
+          List(
+            SqlField("code", country.code, key = true, hidden = true),
+            SqlObject("items", Join(country.code, city.countrycode)),
+            CursorField("offset", genValue("cityOffset"), Nil),
+            CursorField("limit", genValue("cityLimit"), Nil),
+            SqlField("total", country.numCities),
+          )
+      ),
+      ObjectMapping(
+        tpe = CityType,
+        fieldMappings =
+          List(
+            SqlField("id", city.id, key = true, hidden = true),
+            SqlField("countrycode", city.countrycode, hidden = true),
+            SqlField("name", city.name)
+          )
+      )
+    )
+
+  def genValue(key: String)(c: Cursor): Result[Int] =
+    c.env[Int](key).toRightIor(mkOneError(s"Missing key '$key'"))
```

Comment:
FYI there is `envR` which returns a `Result` with a nice error message if the key is missing.

Reply:
Hmm ... I used `envR` elsewhere ... I must have missed this one ... good catch ... I'll tweak it later.